lib/Dialect/ATen/Transforms/RecognizeTorchFallbackPass.cpp

```diff
@@ -0,0 +1,103 @@
+//===- RecognizeTorchFallback.cpp ------------------------------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Dialect/Shape/IR/Shape.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "npcomp/Dialect/ATen/IR/ATenDialect.h"
+#include "npcomp/Dialect/ATen/Transforms/Passes.h"
+#include "npcomp/Dialect/Basicpy/IR/BasicpyDialect.h"
+#include "npcomp/Dialect/Basicpy/IR/BasicpyOps.h"
+#include "npcomp/Dialect/Numpy/IR/NumpyDialect.h"
+#include "npcomp/Dialect/Numpy/IR/NumpyOps.h"
+#include "npcomp/Dialect/Refback/IR/RefbackOps.h"
+#include "npcomp/Dialect/Torch/IR/OpInterfaces.h"
+#include "npcomp/Dialect/Torch/IR/TorchOps.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "aten-recognize-torch-fallback"
+
+using namespace mlir;
+using namespace mlir::NPCOMP;
+using namespace mlir::NPCOMP::aten;
+using namespace mlir::NPCOMP::Basicpy;
+using namespace mlir::NPCOMP::Torch;
+
+namespace {
+
+class EncapsulateKernelCallOpPattern : public OpRewritePattern<Torch::KernelCallOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(Torch::KernelCallOp kernelCall,
+                                PatternRewriter &rewriter) const override {
+    Operation *kcOp = kernelCall.getOperation();
+    // Change nothing if this KernelCallOp does not descend directly from a FuncOp.
+    if (!isa<FuncOp>(kcOp->getParentOp())) {
+      return failure();
+    }
+    // New Torch fallback region.
+    auto torchfb = rewriter.create<refback::TorchFallbackOp>(
```

Comment:
ATen dialect cannot depend on refback. This pass should live in the reference backend.